scheds/rust/scx_flash/src/bpf/main.bpf.c

```diff
@@ -110,322 +342,1422 @@ int kprobe_vfs_fsync_range(struct file *file, u64 start, u64 end, int datasync)
 	tctx = try_lookup_task_ctx(p);
 	if (tctx)
 		tctx->avg_nvcsw = 0;
+
 	return 0;
 }
 
 /*
- * Exponential weighted moving average (EWMA).
- *
- * Copied from scx_lavd. Returns the new average as:
- *
- *	new_avg := (old_avg * .75) + (new_val * .25);
+ * Return true if the target task @p is a kernel thread.
  */
-static u64 calc_avg(u64 old_val, u64 new_val)
+static inline bool is_kthread(const struct task_struct *p)
 {
-	return (old_val - (old_val >> 2)) + (new_val >> 2);
+	return p->flags & PF_KTHREAD;
 }
 
 /*
- * Evaluate the EWMA limited to the range [low ... high]
+ * Return true if @p still wants to run, false otherwise.
  */
-static u64 calc_avg_clamp(u64 old_val, u64 new_val, u64 low, u64 high)
+static bool is_queued(const struct task_struct *p)
 {
-	return CLAMP(calc_avg(old_val, new_val), low, high);
+	return p->scx.flags & SCX_TASK_QUEUED;
 }
 
 /*
- * Return true if the target task @p is a kernel thread, false instead.
+ * Return true if @cpu is in a full-idle physical core,
+ * false otherwise.
  */
-static inline bool is_kthread(const struct task_struct *p)
+static bool is_fully_idle(s32 cpu)
 {
-	return p->flags & PF_KTHREAD;
+	const struct cpumask *idle_smtmask;
+	int node = __COMPAT_scx_bpf_cpu_node(cpu);
+	bool is_idle;
+
+	idle_smtmask = __COMPAT_scx_bpf_get_idle_smtmask_node(node);
+	is_idle = bpf_cpumask_test_cpu(cpu, idle_smtmask);
+	scx_bpf_put_cpumask(idle_smtmask);
+
+	return is_idle;
 }
 
 /*
- * Return the amount of tasks that are waiting to run.
+ * Allocate/re-allocate a new cpumask.
  */
-static inline u64 nr_tasks_waiting(int node)
+static int calloc_cpumask(struct bpf_cpumask **p_cpumask)
```

Comment:
This helper makes me sad to think of how much duplicated code there is for creating cpumasks. Might be nice to throw in `common.bpf.c` as well.

Reply:
I agree. It's now used in flash, bpfland, rustland_core, lavd and tickless. We definitely need to move this to common.bpf.h.